### PR TITLE
[RISCV][GISel] Legalize G_SMULO/G_UMULO

### PR DESCRIPTION
**Pull request by topperc**
I'm unhappy with the XLen case when we need to make a libcall. We get two libcalls.

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -145,13 +145,21 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
         .legalFor({XLenLLT})
         .lower();
     // clang-format on
+
+    getActionDefinitionsBuilder({G_SMULO, G_UMULO})
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:149]: Yeah they're the same. They were briefly different while I tried and failed to fix the 2 libcall issue.

[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:149]: I have a feeling that you are going for coverage and not performance. Could you instead do a `custom` legalization, i.e. RISC-V specific and not generic lower?

[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:149]: The widening of small types creates an XLenLLT sized G_SMULO even though that can't overflow if the small type was half of XLen or less. I don't know how to distinquish that G_SMULO from the original. So I ended up with a double XLen libcall no matter what which is not really what I wanted.

I guess I could write custom legalization for everything, but didn't seem ideal.

**Comment by topperc** [comment]
Ping

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -145,13 +145,31 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
...
 
     getActionDefinitionsBuilder({G_SMULH, G_UMULH}).lowerFor({XLenLLT});
+
+    getActionDefinitionsBuilder({G_SMULO, G_UMULO})
+        .minScalar(0, XLenLLT)
+        // Widen XLenLLT to DoubleXLenLLT so we can use a single libcall to get
+        // the low bits for the mul result and high bits to do the overflow
+        // check.
+        .widenScalarIf(
+            [=, &ST](const LegalityQuery &Query) {
```
Comment (aemerson) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:166]: No need for `ST`?

**Review by aemerson — APPROVED** [review]
LGTM.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:166]: Thanks. Copy and pasted from elsewhere. Will remove.

**Pull request merged**